policy/release/slsa_source_correlated.rego

```diff
@@ -0,0 +1,130 @@
+
+import future.keywords.contains
+import future.keywords.if
+import future.keywords.in
+
+import data.lib
+
+# METADATA
+# title: Expected source code reference not provided
+# description: >-
+#   Warn if the expected source code reference is not provided.
+# custom:
+#   short_name: missing_expected_source_code_reference
+#   failure_msg: Expected source code reference was not provided for verification
+# solution: >-
+#   Provide the expected source code reference in inputs.
+warn contains result if {
+	source := object.get(input, ["image", "source"], {})
+	count(source) == 0
+	result := lib.result_helper(rego.metadata.chain(), [])
+}
+
+# METADATA
+# title: Source reference material not provided
+# description: >-
+#   Attested materials contain source reference.
+# custom:
+#   short_name: material_code_reference
+#   failure_msg: The attested material contains no source code reference
+#   solution: >-
+#     Check that the attestation creation process includes the source code reference
+#     in the predicate.materials for SLSA Provenance v0.2, or in
+#     predicate.buildDefinition.resolvedDependencies for SLSA Provenance v1.0
+#     attestations. Check that the SCM prefix is  the list of the supported SCM
+#     types in rule data (`supported_vcs` key).
+#   collections:
+#   - minimal
+#   - slsa1
+#   - slsa2
+#   - slsa3
+#   - redhat
+#   depends_on:
+#   - attestation_type.known_attestation_type
+#
+deny contains result if {
+	count(_source_references) == 0
+
+	result := lib.result_helper(rego.metadata.chain(), [])
+}
+
+# METADATA
+# title: Provided source code reference is the one that is being attested
+# description: >-
+#   Verify that the provided source code reference is the one being attested.
+# custom:
+#   short_name: expected_source_code_reference
+#   failure_msg: The expected source code reference %q is not the attested one %q
+#   solution: >-
+#     The source code reference in the attestation doesn't match the expected and
+#     provided source code reference. Make sure that the provided source code
+#     reference is correct, and if it is make sure that the build process is
+#     configured to retrieve the source code from the appropriate source code
+#     repository. Make sure that the source code reference is pointing to a
+#     explicit revision not to a symbolic identifier, e.g. a branch or tag name.
+#   collections:
+#   - minimal
+#   - slsa1
+#   - slsa2
+#   - slsa3
+#   - redhat
+#   depends_on:
+#   - attestation_type.known_attestation_type
+#
+deny contains result if {
+	some scm_type in object.keys(input.image.source)
+
+	# e.g. git+https://github.com/...
+	expected_scm_uri := sprintf("%s+%s", [scm_type, input.image.source[scm_type].url])
+	expected_revision := input.image.source[scm_type].revision
+	expected_source := sprintf("%s@%s", [expected_scm_uri, expected_revision])
+
+	some attested_source in _source_references
+
+	expected_source != attested_source
```

Comment:
Instead of an exact match, should this just make sure that one of the items in `_source_references` match? IIUC, resolvedDependencies and materials may include other things. Maybe this?

```rego
expected_source in _source_references
```

Reply:
I don't think so, there is one source reference in `input.image.source`, there should be only one source reference in the attestation. Having more than one violates the expectation of the one source reference from the `input.image.source`.

Reply:
I think Chains will add git+ resources to resolvedDependencies if the git resolver is used: https://github.com/tektoncd/chains/blob/70c8c7de563ba3d6c1e65a1e4e21c83335fa432a/pkg/chains/formats/slsa/v2alpha2/internal/resolved_dependencies/resolved_dependencies_test.go#L519-L547

Reply:
Awkward, brings to light the importance of https://github.com/tektoncd/chains/issues/914, I _think_ I can workaround that by eliminating the references from the steps from the ones considered, the PipelineRun attestation I worked with had a oci bundle IIRC without a resolver IIRC (old school). I'll experiment a bit and see...

Reply:
So, actually not, at least not with Chains v0.17.0 the `resolvedDependencies` contains the images used in the steps, not the git references to the task definitions. See: https://gist.github.com/zregvart/7cc23e92ba37333f70c07af7671cac5a#file-attestation-v1-json-L144-L170

Reply:
We learned that Chains will eventually populate resolvedDependencies with git-resolver repos. We also learned that we don't have a way of distinguishing those from an application source code.

I think we should allow for other git things to be in resolvedDependencies until we have a definitive way of distinguishing them.

Reply:
Using intersection of expected and attested now